crates/wasmtime/src/runtime/vm/traphandlers.rs

```diff
@@ -362,11 +384,32 @@ pub enum TrapReason {
 
     /// A trap raised from a wasm libcall
     Wasm(wasmtime_environ::Trap),
+
+    /// An exception.
+    ///
+    /// Note that internally, exceptions are rooted on the Store, while
+    /// when crossing the public API, exceptions are held in a
+    /// `wasmtime::Exception` which contains a boxed root and implements
+    /// `Error`. This choice is intentional, to keep the internal
+    /// implementation lightweight and ensure the types represent only
+    /// allowable states.
+    #[cfg(feature = "gc")]
+    Exception,
 }
 
 impl From<Error> for TrapReason {
-    fn from(err: Error) -> Self {
-        TrapReason::User(err)
+    #[cfg(feature = "gc")]
+    fn from(error: Error) -> Self {
+        if error.is::<ThrownException>() {
+            TrapReason::Exception
+        } else {
+            TrapReason::User(error)
+        }
+    }
+
+    #[cfg(not(feature = "gc"))]
+    fn from(error: Error) -> Self {
+        TrapReason::User(error)
     }
```

Comment:
FWIW this is an example of where in my subjective opinion `#[cfg]` is overkill. For example if we have `ThrownException`, a unit struct, in the public API when the `gc` feature is disabled that seems fine to me and worth it to avoid growing more `#[cfg]`

Reply:
Hmm, I pulled on this thread a bit, and adding a `ThrownException` unit struct (or uninhabitable enum) allows this to clean up, but then starts pulling a bunch of other things out of `gc`-only config as well (`TrapReason`, more unwind code depending on where we want to draw the boundary...) and honestly it seems cleaner to me to take the sharp line of types-and-enum-arms-don't-exist-for-exceptions in a GC-disabled build? Otherwise there are a bunch of dynamically unreachable failure points, which seems awkward in its own way...